### PR TITLE
solution

### PR DESCRIPTION
**Pull request by abcdesign1**
[DEMO LINK](https://vadym4che.github.io/react_dynamic-list-of-goods/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,27 +1,60 @@
-import React from 'react';
+import React, { useEffect, useState } from 'react';
 import './App.scss';
 import { GoodsList } from './GoodsList';
 
-// import { getAll, get5First, getRed } from './api/goods';
+import { getAll, get5First, getRedGoods } from './api/goods';
+import { Good } from './types/Good';
 // or
 // import * as goodsAPI from './api/goods';
```
Comment (sergii-nosachenko) [inline src/App.tsx:8]: Cleanup comments

File: src/App.tsx [hunk 2]
```diff
@@ -1,27 +1,60 @@
...
 
-    <button type="button" data-cy="all-button">
-      Load all goods
-    </button>
+  useEffect(() => {
 
-    <button type="button" data-cy="first-five-button">
-      Load 5 first goods
-    </button>
+  }, []);
```
Comment (sergii-nosachenko) [inline src/App.tsx:15]: What does this code do?

File: src/App.tsx [hunk 3]
```diff
@@ -1,27 +1,60 @@
...
+    .then((goods) => setGoodsToShow([...goods]));
 
-    <GoodsList goods={[]} />
-  </div>
-);
+  const loadFiveFirstGoods = () => get5First()
+    .then((goods) => setGoodsToShow([...goods]));
+
+  const loadRedGoods = () => getRedGoods()
+    .then((goods) => setGoodsToShow([...goods]));
```
Comment (sergii-nosachenko) [inline src/App.tsx:24]: You need to:
- show loader when fetching data;
- catch and show errors;
- prevent clicking same button twice (don't load same data multiple times in a row)

File: src/api/goods.ts [hunk 4]
```diff
@@ -10,10 +10,12 @@ export function getAll(): Promise<Good[]> {
 
 export const get5First = () => {
   return getAll()
-    .then(goods => goods); // sort and get the first 5
+    .then(goods => goods
+      .sort(({ name: A }, { name: B }) => A.localeCompare(B))
+      .filter((good) => goods.indexOf(good) < 5)); // sort and get the first 5
```
Comment (sergii-nosachenko) [inline src/api/goods.ts:15]: Cleanup comments

**Review by sergii-nosachenko — CHANGES_REQUESTED** [review]

File: src/GoodsList.tsx [hunk 5]
```diff
@@ -7,10 +7,18 @@ type Props = {
 
 export const GoodsList: React.FC<Props> = ({ goods }) => (
   <ul>
-    {goods.map(good => (
-      <li key={good.id} data-cy="good">
-        {good.name}
-      </li>
-    ))}
+    {goods.map((good) => {
+      const { id, color, name } = good;
```
[the remark this answers is not in the document]
Comment (vadym4che) [reply src/GoodsList.tsx:11]: done

File: src/App.tsx [hunk 6]
```diff
@@ -1,27 +1,60 @@
...
+
+  // eslint-disable-next-line no-console
+  console.log(goodsToShow, typeof goodsToShow);
```
[the remark this answers is not in the document]
Comment (vadym4che) [reply src/App.tsx:27]: done

File: src/api/goods.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (vadym4che) [reply src/api/goods.ts:15]: done

File: src/App.tsx (hunk 1 above)
Comment (vadym4che) [reply src/App.tsx:8]: done

File: src/App.tsx (hunk 2 above)
Comment (vadym4che) [reply src/App.tsx:15]: nothing was doing - removed

File: src/api/goods.ts (hunk 4 above)
Comment (vadym4che) [reply src/api/goods.ts:15]: done

File: src/App.tsx (hunk 3 above)
Comment (vadym4che) [reply src/App.tsx:24]: all done

**Review by tiserett — CHANGES_REQUESTED** [review]
good work, but you can fetch your goods in a better  way ;)

File: src/App.tsx [hunk 7]
```diff
@@ -1,27 +1,127 @@
...
+    getAll()
+      .then((goods) => {
+        setGoodsToShow([...goods]);
+        setLoadedState('all');
+        reset();
+      })
+      .catch(() => {
+        setLoadingError(true);
+        reset();
+      });
```
[the remark this answers is not in the document]
Comment (vadym4che) [reply src/App.tsx]: got it
done it
=)

File: src/App.tsx [hunk 8]
```diff
@@ -1,27 +1,119 @@
...
+  };
 
-    <GoodsList goods={[]} />
-  </div>
-);
+  const handleThanAndCatch = (buttonString: Loaded) => {
+    return () => {
+      // This code uses an immediately invoked function expression (IIFE)
+      // to execute the switch statement and
+      // return the appropriate function based on the value of `buttonString`.
```
Comment (sergii-nosachenko) [inline src/App.tsx:32]:
```suggestion
```

File: src/App.tsx [hunk 9]
```diff
@@ -1,27 +1,119 @@
...
+          case 'all':
+            return getAll;
+          case 'five':
+            return get5First;
+          case 'red':
+            return getRedGoods;
+          default:
+            return getAll;
+        }
+      })(); // The returned function is then assigned to `getterFunction` using const.
```
Comment (sergii-nosachenko) [inline src/App.tsx:44]:
```suggestion
      })();
```

File: src/App.tsx [hunk 10]
```diff
@@ -1,27 +1,119 @@
...
-    <h1>Dynamic list of Goods</h1>
+export const App: React.FC = () => {
+  const [goodsToShow, setGoodsToShow] = useState<Good[]>([]);
+  const [loading, setLoading] = useState(false);
+  const [loadingError, setLoadingError] = useState(false);
 
-    <button type="button" data-cy="all-button">
-      Load all goods
-    </button>
+  type Loaded = 'all' | 'five' | 'red' | 'none';
```
Comment (sergii-nosachenko) [inline src/App.tsx:18]: Enum is better for such purpose

File: src/components/GoodsList/GoodsList.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+}) => {
+  return (
+    <>
+      {loadingError
+        ? (
+          <div className="notification is-danger is-light error">
+            {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
+            <button
+              className="delete"
+              type="button"
```
Comment (sergii-nosachenko) [inline src/components/GoodsList/GoodsList.tsx:23]:
```suggestion
            <button
              className="delete"
              aria-label="Close notification"
              type="button"
```

**Review by sergii-nosachenko — APPROVED** [review]

**Pull request closed without merging**